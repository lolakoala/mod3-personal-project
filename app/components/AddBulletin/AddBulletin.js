import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AddBulletin extends Component {
  constructor() {
    super();
    this.state = {
      title: '',
      details: ''
    };
  }

  handleChange(event, type) {
    const { value } = event.target;
    this.setState({
      [type]: value
    });
  }

  addBulletin = () => {
    const bulletin = {
      id: Date.now(),
      title: this.state.title,
      details: this.state.details,
      hasRead: [this.props.currentUser.id],
      postedBy: this.props.currentUser,
      datePosted: this.props.getTodaysDate()
    };

    this.props.addBulletin(bulletin, this.props.usersHouse);
    this.props.updateItem('');
  }

  render() {
    return (
      <div>
        <input type='text' placeholder='Title' onChange={(event) => this.handleChange(event, 'title')}/>
        <textarea type='text' placeholder='Details' onChange={(event) => this.handleChange(event, 'details')}/>
        <button onClick={() => this.addBulletin()}>Submit</button>
      </div>
    );
  }
}

export default AddBulletin;

AddBulletin.propTypes = {
  usersHouse: PropTypes.object,
  currentUser: PropTypes.object,
  addBulletin: PropTypes.func,
  getTodaysDate: PropTypes.func,
  updateItem: PropTypes.func
};
