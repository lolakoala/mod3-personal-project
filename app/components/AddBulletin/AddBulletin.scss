.add-bulletin input,
.add-bulletin button,
.add-bulletin textarea {
  font-family: 'BenchNine', sans-serif;
  border-radius: 25px;
  margin-top: 20px;
  height: 25px;
  font-size: 20px;
  width: 120px;
  margin-left: 10px;
  margin-right: 10px;
}

.add-bulletin input,
.add-bulletin textarea {
  padding-left: 10px;
  border: 1px solid #D83D59;
  color: #D83D59;
  border-radius: 5px;
}

.add-bulletin input::placeholder,
.add-bulletin textarea::placeholder {
  color: #D83D59;
}

.add-bulletin button {
  border: none;
  background-color: #D83D59;
  color: #F4CCE2;
}

.add-bulletin button:hover,
.add-bulletin .active {
  cursor: pointer;
  border: 2px dashed #D83D59;
  background-color: #F4CCE2;
  color: #D83D59;
}

.add-bulletin textarea {
  height: 70px;
}

.add-bulletin-inputs,
.add-bulletin-buttons {
  display: flex;
  flex-direction: column;
}

.add-bulletin{
  display: flex;
  margin-left: 20px;
}
