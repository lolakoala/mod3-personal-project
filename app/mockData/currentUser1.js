const currentUser1 = {
  id : "RVde1m9BhkQBLZlTJNsFlQZXqP43",
  name: "Amanda Brenner",
  house: "Dog9"
};

export default currentUser1;
