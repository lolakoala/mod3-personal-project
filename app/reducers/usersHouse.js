import firebase from '../firebase.js';

const usersHouse = (state = {}, action) => {
  const housesRef = firebase.database().ref('houses');
  switch (action.type) {
  case 'CREATE_HOUSE':
    housesRef.push(action.house);
    return action.house;
  case 'GET_HOUSE':
    const updatedHouse = Object.assign(action.usersHouse, { users: [...action.usersHouse.users, action.user] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updatedHouse);
    return updatedHouse;
  case 'LOGIN_SUCCESS':
    return action.usersHouse;
  case 'SIGNOUT':
    return {};
  case 'LEAVE_HOUSE':
    const deleteUser = action.usersHouse.users.filter(user => user.id !== action.currentUser.id);
    const houseWithoutUser = Object.assign(action.usersHouse, { users: deleteUser });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithoutUser);
    return {};
  case 'ADD_BILL':
    const takeOutFakeBills = action.usersHouse.bills.filter(bill => bill.title !== 'fake');
    const houseWithBill = Object.assign(action.usersHouse, { bills: [...takeOutFakeBills, action.bill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBill);
    return houseWithBill;
  case 'ADD_BULLETIN':
    const takeOutFakeBulletins = action.usersHouse.bulletins.filter(bulletin => bulletin.title !== 'fake');
    const houseWithBulletin = Object.assign(action.usersHouse, { bulletins: [...takeOutFakeBulletins, action.bulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBulletin);
    return houseWithBulletin;
  case 'ADD_CHORE':
    const takeOutFakeChore = action.usersHouse.chores.filter(chore => chore.title !== 'fake');
    const houseWithChore = Object.assign(action.usersHouse, { chores: [...takeOutFakeChore, action.chore] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithChore);
    return houseWithChore;
  case 'ADD_READER':
    const oldBulletin = action.usersHouse.bulletins.find(bulletin => bulletin.id === action.bulletinId);
    if (oldBulletin.hasRead.includes(action.userId)) {
      return action.usersHouse;
    }
    const newBulletin = Object.assign(oldBulletin, { hasRead: [...oldBulletin.hasRead, action.userId] });
    const updateBulletins = action.usersHouse.bulletins.filter(bulletin => {
      return bulletin.id !== action.bulletinId;
    });
    const updateWithBulletin = Object.assign(action.usersHouse, { bulletins: [...updateBulletins, newBulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBulletin);
    return updateWithBulletin;
  case 'MARK_PAID':
    const oldBill = action.usersHouse.bills.find(bill => bill.id === action.billId);
    const user = oldBill.allUsersTotals.find(user => user.id === action.userId);
    if (user.paid === true) {
      return action.usersHouse;
    }
    const newUser = Object.assign(user, { paid: true });
    const newAllUsersTotals = oldBill.allUsersTotals.filter(user => user.id !== action.userId);
    const newBill = Object.assign(oldBill, { allUsersTotals: [...newAllUsersTotals, newUser] });
    const updatedBills = action.usersHouse.bills.filter(bill => {
      return bill.id !== action.billId;
    });
    const updateWithBill = Object.assign(action.usersHouse, { bills: [...updatedBills, newBill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBill);
    return updateWithBill;
  default:
    return state;
  }
};

export default usersHouse;
