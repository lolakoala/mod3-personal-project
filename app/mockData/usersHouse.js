import currentUser1 from './currentUser1.js';
import currentUser2 from './currentUser2.js';

const bill1 = {
  allUsersTotals: [
    {
      id: currentUser1.id,
      paid: false,
      total: 50
    },
    {
      id: currentUser2.id,
      paid: false,
      total: 50
    }
  ],
  datePosted: "11/01/2017",
  details: "We used a lot of water.",
  duedate: "11/9/17",
  id: "b3019762705779",
  parsedDuedate: 1510210800000,
  postedBy: currentUser1,
  title: "Water Bill",
  total: "100"
};

const bulletin1 = {
  datePosted: "11/01/2017",
  details: "The dogs got out through a hole in the fence yesterday.",
  hasRead: [currentUser1.id],
  id: 1509567098747,
  postedBy: currentUser1,
  title: "Don't leave dogs out unattended."
};

const chore1 = {
  assignedTo: currentUser1.id,
  datePosted: "10/31/2017",
  details: "Empty the dishwasher.",
  done: false,
  id: "c1509485241076",
  postedBy: currentUser1,
  title: "Dishes",
  urgency: "low"
};

const usersHouse = {
  houseKey: "-KxoIBWy5UYRscFYdWty",
  bills: [bill1],
  bulletins: [bulletin1],
  chores: [chore1],
  houseCode: "woof9",
  houseName: "Dog9",
  users: [currentUser1, currentUser2]
};

export default usersHouse;
