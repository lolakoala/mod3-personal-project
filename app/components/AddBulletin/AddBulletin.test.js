import React from "react";
import AddBulletin from './AddBulletin.js';
import { shallow } from "enzyme";
import currentUser1 from '../../mockData/currentUser1.js';
import usersHouse from '../../mockData/usersHouse.js';

describe('AddBulletin component', () => {
  const mockFn = jest.fn();
  const wrapper = shallow(<AddBulletin usersHouse={usersHouse}
    currentUser={currentUser1}
    addBulletin={mockFn}
    getTodaysDate={mockFn}
    updateItem={mockFn}/>);
  const input1 = wrapper.find('input');
  const input2 = wrapper.find('textarea');

  it('should exist', () => {
    expect(wrapper).toBeDefined();
  });

  it('should have a default state', () => {
    const defaultState = {
      title: '',
      details: '',
      submitDisabled: true
    };

    expect(wrapper.state()).toEqual(defaultState);
  });

  it('should render 2 inputs and 2 buttons', () => {
    const buttons = wrapper.find('button');

    expect(input1.length + input2.length).toEqual(2);
    expect(buttons.length).toEqual(2);
  });

  it('should change state on input', () => {
    const newState = {
      title: 'Gas Bill Due',
      details: 'Pay Gas Bill Today',
      submitDisabled: false
    };

    input1.simulate('change', { target: { value: 'Gas Bill Due' } });
    input2.simulate('change', { target: { value: 'Pay Gas Bill Today' } });
    expect(wrapper.state()).toEqual(newState);
  });

  it('should fire actions', () => {
    const submit = wrapper.find('button').first();

    submit.simulate('click');
    expect(mockFn).toHaveBeenCalledTimes(3);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
