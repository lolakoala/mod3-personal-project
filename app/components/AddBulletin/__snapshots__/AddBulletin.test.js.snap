// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBulletin component should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddBulletin
    addBulletin={[Function]}
    currentUser={
      Object {
        "house": "Dog9",
        "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
        "name": "Amanda Brenner",
      }
    }
    getTodaysDate={[Function]}
    updateItem={[Function]}
    usersHouse={
      Object {
        "bills": Array [
          Object {
            "allUsersTotals": Array [
              Object {
                "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
                "paid": false,
                "total": 50,
              },
              Object {
                "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
                "paid": false,
                "total": 50,
              },
            ],
            "datePosted": "11/01/2017",
            "details": "We used a lot of water.",
            "duedate": "11/9/17",
            "id": "b3019762705779",
            "parsedDuedate": 1510210800000,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Water Bill",
            "total": "100",
          },
        ],
        "bulletins": Array [
          Object {
            "datePosted": "11/01/2017",
            "details": "The dogs got out through a hole in the fence yesterday.",
            "hasRead": Array [
              "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            ],
            "id": 1509567098747,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Don't leave dogs out unattended.",
          },
        ],
        "chores": Array [
          Object {
            "assignedTo": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "datePosted": "10/31/2017",
            "details": "Empty the dishwasher.",
            "done": false,
            "id": "c1509485241076",
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Dishes",
            "urgency": "low",
          },
        ],
        "houseCode": "woof9",
        "houseKey": "-KxoIBWy5UYRscFYdWty",
        "houseName": "Dog9",
        "users": Array [
          Object {
            "house": "Dog9",
            "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "name": "Amanda Brenner",
          },
          Object {
            "house": "Dog9",
            "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
            "name": "Lola Koala",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="add-bulletin-inputs"
        >
          <input
            className="bulletin-input"
            onChange={[Function]}
            placeholder="Title"
            type="text"
          />
          <textarea
            className="bulletin-input"
            onChange={[Function]}
            placeholder="Details"
            type="text"
          />
        </div>,
        <div
          className="add-bulletin-buttons"
        >
          <button
            disabled={false}
            onClick={[Function]}
          >
            Submit
          </button>
          <button
            onClick={[Function]}
          >
            Reset
          </button>
        </div>,
      ],
      "className": "add-bulletin",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="bulletin-input"
              onChange={[Function]}
              placeholder="Title"
              type="text"
            />,
            <textarea
              className="bulletin-input"
              onChange={[Function]}
              placeholder="Details"
              type="text"
            />,
          ],
          "className": "add-bulletin-inputs",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "className": "bulletin-input",
              "onChange": [Function],
              "placeholder": "Title",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "className": "bulletin-input",
              "onChange": [Function],
              "placeholder": "Details",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "textarea",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              disabled={false}
              onClick={[Function]}
            >
              Submit
            </button>,
            <button
              onClick={[Function]}
            >
              Reset
            </button>,
          ],
          "className": "add-bulletin-buttons",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "disabled": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Reset",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Reset",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="add-bulletin-inputs"
          >
            <input
              className="bulletin-input"
              onChange={[Function]}
              placeholder="Title"
              type="text"
            />
            <textarea
              className="bulletin-input"
              onChange={[Function]}
              placeholder="Details"
              type="text"
            />
          </div>,
          <div
            className="add-bulletin-buttons"
          >
            <button
              disabled={false}
              onClick={[Function]}
            >
              Submit
            </button>
            <button
              onClick={[Function]}
            >
              Reset
            </button>
          </div>,
        ],
        "className": "add-bulletin",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="bulletin-input"
                onChange={[Function]}
                placeholder="Title"
                type="text"
              />,
              <textarea
                className="bulletin-input"
                onChange={[Function]}
                placeholder="Details"
                type="text"
              />,
            ],
            "className": "add-bulletin-inputs",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "className": "bulletin-input",
                "onChange": [Function],
                "placeholder": "Title",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "className": "bulletin-input",
                "onChange": [Function],
                "placeholder": "Details",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "textarea",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                disabled={false}
                onClick={[Function]}
              >
                Submit
              </button>,
              <button
                onClick={[Function]}
              >
                Reset
              </button>,
            ],
            "className": "add-bulletin-buttons",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Submit",
                "disabled": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Submit",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Reset",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Reset",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
